File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import heapq
import sys

input = sys.stdin.readline
V, E = map(int, input().split())

visited = [False] * (V + 1)
Elist = [[] for _ in range(V + 1)]
heap = [[0, 1]]

for i in range(E):
    s, e, w = map(int, input().split())
    Elist[s].append([w, e])
    Elist[e].append([w, s])

answer, cnt = 0, 0

while heap:
    if cnt == V:
        break
    w, s = heapq.heappop(heap)
    if not visited[s]:
        visited[s] = True
        answer += w
        cnt += 1
        for i in Elist[s]:
            heapq.heappush(heap, i)
print(answer)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.cc
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  int V, E;
  cin >> V >> E;

  vector<bool> visited(V + 1, false);
  vector<vector<pair<int, int>>> Elist(V + 1);
  priority_queue<pair<int, int>, vector<pair<int, int>>,
                 greater<pair<int, int>>>
      heap;
  heap.push(make_pair(0, 1));

  for (int i = 0; i < E; ++i) {
    int s, e, w;
    cin >> s >> e >> w;
    Elist[s].emplace_back(w, e);
    Elist[e].emplace_back(w, s);
  }

  int answer = 0, cnt = 0;

  while (!heap.empty()) {
    if (cnt == V) {
      break;
    }
    int w = heap.top().first;
    int s = heap.top().second;
    heap.pop();
    if (!visited[s]) {
      visited[s] = true;
      answer += w;
      ++cnt;
      for (const auto &i : Elist[s]) {
        heap.push(i);
      }
    }
  }

  cout << answer << endl;

  return 0;
}